eradiate/scenes/biosphere/_discrete.py

```diff
@@ -1125,6 +1117,11 @@ def shapes(self, ctx=None):
         kernel_height = self.trunk_height.m_as(kernel_length)
         kernel_radius = self.trunk_radius.m_as(kernel_length)
 
+        # offset the leaf cloud by the size of the trunk
+        self.leaf_cloud.leaf_positions += (
+            np.array([0, 0, kernel_height]).reshape((1, 3)) * kernel_length
+        )
```

Comment:
Danger-danger: you're mutating the object and you shouldn't.
```suggestion
        leaf_positions = self.leaf_cloud.leaf_positions + np.array([0, 0, kernel_height]).reshape((1, 3)) * kernel_length
```

Reply:
I don't see another way to get the offset information into the `leaf_cloud` object. I don't think it is a good idea to add the offset as a parameter to the `kernel_dict` method.
If anything, the leaf cloud can get a new member variable again, which holds the offset and which is then written into by `AbstractTree` :man_shrugging: 
Or do you think I should modify all the transformation matrices after I called `leaf_cloud.shape()` ?

Reply:
Also, using `+=` does not work here anyways. (ref: https://techoverflow.net/2019/05/22/how-to-fix-numpy-typeerror-cannot-cast-ufunc-subtract-output-from-dtypefloat64-to-dtypeint64-with-casting-rule-same_kind/)

Reply:
> I don't see another way to get the offset information into the leaf_cloud object. I don't think it is a good idea to add the offset as a parameter to the kernel_dict method.

The context system should handle this kind of cases. The best is probably to subclass `KernelDictContext` (*e.g.* as a local member of `LeafCloud`), add a `translate` field to it which will take a 3-vector as value and initialise an instance of this from a parent `KernelDictContext`.

> Also, using += does not work here anyways. (ref: https://techoverflow.net/2019/05/22/how-to-fix-numpy-typeerror-cannot-cast-ufunc-subtract-output-from-dtypefloat64-to-dtypeint64-with-casting-rule-same_kind/)

I think it would with `np.array([0, 0, kernel_height], dtype=float).reshape((1, 3))`.